File: class23/java/PES2301-ComandosGit.java
/*
 * PES2301 - Roteiro de Comandos Git Comentado
 * Aula 23: Gerência de Configuração e Introdução ao Git
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Compilar/executar:
 *   javac PES2301-ComandosGit.java && java ComandosGit
 */

class ComandosGit {

    static class ComandoGit {
        String comando, descricao, categoria;

        ComandoGit(String comando, String descricao, String categoria) {
            this.comando = comando; this.descricao = descricao; this.categoria = categoria;
        }
    }

    static final ComandoGit[] ROTEIRO = {
        new ComandoGit("git init", "Inicializa um novo repositório Git", "inicialização"),
        new ComandoGit("git clone <url>", "Clona um repositório remoto", "inicialização"),
        new ComandoGit("git status", "Mostra o estado dos arquivos", "consulta"),
        new ComandoGit("git add <arquivo>", "Adiciona à staging area", "staging"),
        new ComandoGit("git add .", "Adiciona todos os modificados", "staging"),
        new ComandoGit("git commit -m 'msg'", "Registra mudanças no repo local", "commit"),
        new ComandoGit("git log --oneline", "Histórico de commits resumido", "consulta"),
        new ComandoGit("git diff", "Diferenças WD vs staging", "consulta"),
        new ComandoGit("git push origin main", "Envia commits para remoto", "remoto"),
        new ComandoGit("git pull origin main", "Baixa e mescla do remoto", "remoto"),
        new ComandoGit("git branch <nome>", "Cria uma nova branch", "branch"),
        new ComandoGit("git checkout <branch>", "Muda para a branch", "branch"),
        new ComandoGit("git merge <branch>", "Mescla branch na atual", "branch"),
    };

    public static void main(String[] args) {
        System.out.println("=".repeat(70));
        System.out.println("  GUIA RÁPIDO DE COMANDOS GIT");
        System.out.println("=".repeat(70));
        System.out.println();

        String categoriaAtual = "";
        for (ComandoGit c : ROTEIRO) {
            if (!c.categoria.equals(categoriaAtual)) {
                categoriaAtual = c.categoria;
                System.out.println("--- " + categoriaAtual.toUpperCase() + " ---");
            }
            System.out.printf("  $ %-32s # %s%n", c.comando, c.descricao);
        }

        System.out.println("\n--- BOAS PRÁTICAS: MENSAGENS DE COMMIT ---");
        System.out.println("  ✓ feat: adiciona endpoint de login");
        System.out.println("  ✓ fix: corrige validação de email nulo");
        System.out.println("  ✗ arrumei umas coisas (vago)");
        System.out.println("  ✗ atualizacao (genérico)");
    }
}

File: class23/python/PES2301-ComandosGit.py
"""
PES2301 - Roteiro de Comandos Git Comentado
Aula 23: Gerência de Configuração e Introdução ao Git
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Exibe um roteiro comentado dos principais comandos Git,
funcionando como referência rápida para o laboratório prático.
"""

ROTEIRO_GIT = [
    # (comando, descrição, categoria)
    ("git init", "Inicializa um novo repositório Git no diretório atual", "inicialização"),
    ("git clone <url>", "Clona um repositório remoto para o diretório local", "inicialização"),
    ("git status", "Mostra o estado dos arquivos (modificados, staged, untracked)", "consulta"),
    ("git add <arquivo>", "Adiciona arquivo à staging area (prepara para commit)", "staging"),
    ("git add .", "Adiciona todos os arquivos modificados à staging area", "staging"),
    ("git commit -m 'mensagem'", "Registra as mudanças no repositório local", "commit"),
    ("git log --oneline", "Exibe o histórico de commits resumido", "consulta"),
    ("git diff", "Mostra as diferenças entre working directory e staging", "consulta"),
    ("git push origin main", "Envia commits locais para o repositório remoto", "remoto"),
    ("git pull origin main", "Baixa e mescla mudanças do repositório remoto", "remoto"),
    ("git branch <nome>", "Cria uma nova branch", "branch"),
    ("git checkout <branch>", "Muda para a branch especificada", "branch"),
    ("git merge <branch>", "Mescla a branch especificada na branch atual", "branch"),
    ("git status --ignored", "Mostra também arquivos ignorados pelo .gitignore", "consulta"),
    (".gitignore", "Arquivo que lista padrões de arquivos a serem ignorados", "config"),
]


def exibir_roteiro():
    print("=" * 70)
    print("  GUIA RÁPIDO DE COMANDOS GIT")
    print("=" * 70)
    print()

    categorias = {}
    for cmd, desc, cat in ROTEIRO_GIT:
        categorias.setdefault(cat, []).append((cmd, desc))

    for cat, comandos in categorias.items():
        print(f"--- {cat.upper()} ---")
        for cmd, desc in comandos:
            print(f"  $ {cmd:<32} # {desc}")
        print()


def verificar_boas_praticas_commit():
    """Exibe boas práticas para mensagens de commit."""
    print("--- BOAS PRÁTICAS: MENSAGENS DE COMMIT ---")
    boas = [
        ("✓ feat: adiciona endpoint de login", "Prefixo semântico + descrição clara"),
        ("✓ fix: corrige validação de email nulo", "Indica correção de bug"),
        ("✗ arrumei umas coisas", "Vago — não descreve o que foi feito"),
        ("✗ atualizacao", "Genérico — impossível rastrear mudanças"),
    ]
    for msg, explicacao in boas:
        print(f"  {msg:<50} ({explicacao})")


if __name__ == "__main__":
    exibir_roteiro()
    verificar_boas_praticas_commit()
